src/d3d12/nvapi_d3d12_device.cpp

```diff
@@ -80,4 +80,17 @@ namespace dxvk {
 
         return deviceExt;
     }
+
+    Com<ID3D12GraphicsCommandListExt> NvapiD3d12Device::GetCommandListExt(ID3D12GraphicsCommandList* commandList) {
+        auto it = m_CommandListMap.find(commandList);
```

Comment:
Shouldn't this be protected by a mutex or something like that?

Reply:
Good point! Currently all NvAPIs are assuming they are called in thread-safe way, since all other APIs are implemented this way.

Reply:
I think it's technically necessary for the D3D11 NvAPI functions too because you could call it from different threads with different DeferredContexts. Access to the context itself isn't expected to be thread safe, so I think the global map is the only problematic part.

Reply:
Unless I'm missing something, that method looks now different, there is no longer a map for CommandLists, see https://github.com/jp7677/dxvk-nvapi/blob/master/src/d3d12/nvapi_d3d12_device.cpp#L38

For other places where the cubinDevice is looked up in the map (and also for similar maps in nvapi D3D11), from my understanding, worst thing that might happen in a concurrent scenario is that `emplace` is called twice for the same device, which should be ok.

Reply:
> worst thing that might happen in a concurrent scenario is that emplace is called twice for the same device, which should be ok.

Is that okay though? Would rehashing for example not be a problem in that case?

Even if it's okay for devices, it needs to be handled for CommandLists and for DeviceContexts on the D3D11 side.

Reply:
I was searching for the documentation in NvAPI for multi-threading, and could not find, so it looks like we need to add locks here. @K0bin thanks for the catch!
@jp7677 I would create separate PR for locks for others.

Reply:
> > worst thing that might happen in a concurrent scenario is that emplace is called twice for the same commandList,
 
> No, games create multiple command lists and record those on different threads. They could certainly call into NvAPI from those worker threads simultanously with different command lists. (As usual, same thing on the D3D11 side just with DeferredContexts)


From my understanding / how it was intended: The map here just acts like a cache, to prevent lots of `QueryInterface` calls. Having concurrent calls with different `commandList` would  just add multiple entries to the map and not interfere with each other. Having concurrent calls with the same `commandList` might result in a cache-miss, thus for both calls we would do the QueryInterface calls for the same `commandList` and then try to insert the same device into the map with `emplace`. From my understanding how emplace works, even when it is called twice for the same key, both calls should succeed (with the key just added once effectively.). Did I missed something?

Reply:
PS: Just to be sure, I really appreciate the feedback, changes are indeed very real that I had missed something, though I want to understand the issue.

Reply:
I don't think it's guaranteed that concurrent emplace with different keys is safe. It either needs to do rehashing or add the element to a per-bucket list. Neither of which should be thread safe.

Reply:
This needs a mutex.

Reply:
@rochaudhari thanks for the updated PR, alright, convinced we need a mutex for `emplace` :). Not sure if this works elegantly, but could you try to place the lock for the `emplace` call only and not for everything in this method? (My experience is a bit limited in that area in c++.)

Reply:
> but could you try to place the lock for the emplace call only and not for everything in this method? (My experience is a bit limited in that area in c++.)

That could lead to reads while you're writing which would also break.

Reply:
Break in the sense of crash or just no result for find()?

Reply:
(Having the lock in the hot path might defeat the whole cashing thing, that’s why I’m asking)

Reply:
I think it's undefined behavior, so crashes are possible. You could use a Read-Write lock (concurrent reads are supported).

FWIW, Direct3D9, 10 and 11 have an option to use an internal mutex for every single context function and DXVK just uses a spin lock for that.

Reply:
Alright, after having a short chat on discord, seems a shared_ mutex is the way to like described here https://stackoverflow.com/questions/38061067/data-race-with-stdunordered-map-despite-locking-insertions-with-mutex

Reply:
Thanks for the link. I agree that read write lock is more appropriate however, I also fond this: https://stackoverflow.com/questions/50972345/when-is-stdshared-timed-mutex-slower-than-stdmutex-and-when-not-to-use-it

Which also states that: "When the read operation is only a lookup in an in-memory data structure, most likely a simple lock will outperform the read/writer solution." 
Since this locks also comes with little more cost. 
What do you think?

Reply:
Thanks, this topic really isn't the easiest one ;) Following your links, a simple mutex is indeed better.
Since I prefer simple solutions, lets do just this (actually you already did) until we face any issues in real life.

Reply:
PS: `nvapi.cpp` we use `scoped_lock`, we might wanna use that (mostly for consistency). From what I read, there is no penalty for performance and we (currently) don't need the extra capabilities of `unique_lock`

Reply:
It was a good learning exercise :)